Readme.md

```diff
@@ -0,0 +1,1432 @@
+
+```scala
+import scala.collection.JavaConversions._
+
+val s: Vector[Int] = Vector(1, 2, 3)
+val x: java.util.List[Int] = s  // gets converted automatically
+```
+
+Using `JavaConversions` (and additionally `wrapAsJava`, `wrapAsScala` and `wrapAll`) is strictly forbidden; in case if such conversions are necessary, `decoratedAsJava`, `decorateAsScala` or `decorateAll` should be used:
+
+```scala
+import scala.collection.convert.decorateAsJava._
+
+val s: Vector[Int] = Vector(1, 2, 3)
+val x: java.util.List[Int] = s.asJava
+```
+
+### The expression problem, extension methods
+
+[Expression problem](https://en.wikipedia.org/wiki/Expression_problem) is basically a task of adding new operations to an already defined type. It is well known that most of the "old" static languages, like C, C++ and Java, do not have any kind of solution for it. In most dynamic languages this is not a problem due to dynamic typing and monkey patching, but static languages also have their own solutions. For example, C# and Kotlin have extension methods, while Haskell and Rust have type classes (called traits in the latter). Scala solution for the expression problem are methods injected through implicits.
+
+Modern IDEs, in particular, IntelliJ IDEA, are able to understand most if not all commonly used implicits usage scenarios: in particular, they are able to highlight methods added through implicit classes and show implicit parameters of a method call, and allow to navigate to the definition of these methods. Therefore, using implicits for simple tasks, like adding new methods to existing types, is not forbidden:
+
+```scala
+implicit class ArrayOps[A](val arr: Array[A]) extends AnyVal {
+  def firstTwoOpt: Option[(A, A)] = if (arr.length >= 2) Some((arr(0), arr(1))) else None
+}
+```
+
+Note, however, that for this to be acceptable, such extension must be motivated by a significant increase of readability. In particular, this is justified when the newly added operation is used very often in some part of code. If this operation is defined as a regular function, it would require calling it with the function syntax instead of the method syntax:
+
+```scala
+arr.firstTwoOpt
+// vs
+firstTwoOpt(arr)
+```
+
+The function syntax requires parentheses, and when there are many calls of this function or if they are present in some complex expression which already has parentheses, it hurts readability. Extension methods, therefore, may be used if they help to reduce the noise.
+
+Extension methods must only be defined with an `implicit class` declaration, never via an `implicit def` conversion (see the above section). If possible, such `implicit class` must be defined in some object (including package objects) and extend `AnyVal` in order to avoid performance penalties of boxing.  It makes sense to add such operations to some common internal utilities library, and putting them *not* there should be avoided.
```

Comment:
I think I have a vague understanding of what you mean but if this understanding is correct then I would argue it is a very bad practice. It is actually a different kind of an implicit conversion, as far as I can see. I also don't think I ever saw such a thing being used. For handling such use case I would suggest writing a conversion function which wraps the existing class in the new interface and *maybe* expose it as an extension method.